balancer/base/balancer.go

```diff
@@ -101,17 +101,40 @@ func (b *baseBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
 	// addrsSet is the set converted from addrs, it's used for quick lookup of an address.
 	addrsSet := make(map[resolver.Address]struct{})
 	for _, a := range s.ResolverState.Addresses {
-		addrsSet[a] = struct{}{}
-		if _, ok := b.subConns[a]; !ok {
+		// Strip attributes from addresses before using them as map keys. So
+		// that when two addresses only differ in attributes pointers (but with
+		// the same attribute content), they are considered the same address.
+		//
+		// Note that this doesn't handle the case where the attribute content is
+		// different. So if users want to set different attributes to create
+		// duplicate connections to the same backend, it doesn't work. This is
+		// fine for now, because duplicate is done by setting Metadata today.
+		//
+		// TODO: read attributes to handle duplicate connections.
+		aNoAttrs := a
+		aNoAttrs.Attributes = nil
+		addrsSet[aNoAttrs] = struct{}{}
+		if sc, ok := b.subConns[aNoAttrs]; !ok {
 			// a is a new address (not existing in b.subConns).
+			//
+			// When creating SubConn, the original address with attributes is
+			// passed through. So that connection configurations in attributes
+			// (like creds) will be used.
 			sc, err := b.cc.NewSubConn([]resolver.Address{a}, balancer.NewSubConnOptions{HealthCheckEnabled: b.config.HealthCheck})
 			if err != nil {
 				logger.Warningf("base.baseBalancer: failed to create new SubConn: %v", err)
 				continue
 			}
-			b.subConns[a] = sc
+			b.subConns[aNoAttrs] = sc
 			b.scStates[sc] = connectivity.Idle
 			sc.Connect()
+		} else {
+			// Always update the subconn's address in case the attributes
+			// changed. This is a noop if the current address is the same as the
+			// old one (reflect.DeepEqual).
```

Comment:
What does this mean?  It shouldn't be a "true" nop.  It should make it so these attributes are used if we ever need to reconnect.

Of course, we don't actually have any attributes that matter right now, besides the network type, which isn't something that it makes sense to change over time for the same address.

Maybe we should update the grpclb usage of `Metadata` for the server name to go through attributes instead so we can test this?

Reply:
If reflect.DeepEqual returns true, `sc.UpdateAddresses` would immediately return.
reflect.DeepEqual also compares the attributes, so this early return only happens when the new addresses is exactly the same as old.

This change has nothing to do with grpclb. And we will need changes in ClientConn to handle attributes.
That includes making the attributes hashable (or comparable?), and distinguish between the attributes that matter (for connection, e.g. creds), and those that only for balancers to use (e.g. hierarchy)

Reply:
> reflect.DeepEqual also compares the attributes

Oh, I see what you are saying here.  My mistake.

> This change has nothing to do with grpclb.

No, but that's an example of a potential attribute that could affect the transport, besides network type.  Changing it would allow us to test the change you just made.  Actually, we could also implement a dummy handshaker to read the attributes and verify they were updated when a reconnect happens.

What I was thinking was:

1. Push an address from the resolver with no attributes
2. Connect
3. Push the same address from the resolver but with new attributes
4. Force disconnect/reconnect
5. Verify new attributes are in use

Reply:
Added a test